### PR TITLE
feat(usage): ingest external usage events for API-equivalent cost tracking

### LINKED ISSUE #378: feat: ingest external usage events for API-equivalent cost tracking
> _Drafted by Claude through back-and-forth with @njbrake. The direction and prioritization are his; the prose is Claude's._

# feat: ingest external usage events for API-equivalent cost tracking

## Problem

Otari can only track requests that it routes itself. That works for API-backed clients, but it excludes subscription-backed coding agents such as Claude Code.

A Claude Code subscription can expose precise per-request usage locally and through its supported OpenTelemetry export: model, timestamp, input/output tokens, prompt-cache read/write tokens, request ID, duration, and an estimated API-equivalent cost. Routing Claude Code through Otari instead would use an API key and incur API billing, defeating the purpose of retaining the subscription.

Users therefore cannot answer a useful question in Otari today: "What would my subscription-backed Claude Code usage have cost at API rates?"

This is not just a missing connector:

- `UsageLog` has no source, external event ID, or external attribution fields.
- `/v1/usage` is read-only.
- Existing usage rows are assumed to have come from a request Otari served.
- Imported usage must not mutate `User.spend`, reserve a budget, or be used for enforcement, because it is retrospective accounting.
- Hybrid mode has no local usage database or management surface.

## Proposed solution

Add a standalone-only, master-key-protected external usage ingestion API. It accepts normalized, content-free usage events from local agents, collectors, or importers and stores them alongside gateway usage while clearly identifying their source and estimated-cost status.

Initial target: Claude Code's supported OpenTelemetry `api_request` events, sent through an OpenTelemetry Collector or lightweight local adapter.

### API

Add:

```http
POST /v1/usage/external-events
Otari-Key: <master key>
Content-Type: application/json
```

Example request:

```json
{
  "source": "claude_code",
  "events": [
    {
      "source_event_id": "req_01ABC...",
      "timestamp": "2026-07-22T12:34:56Z",
      "provider": "anthropic",
      "model": "claude-sonnet-4-6",
      "status": "success",
      "input_tokens": 1200,
      "output_tokens": 450,
      "cache_read_tokens": 8000,
      "cache_write_tokens": 1024,
      "duration_ms": 4821,
      "session_label": "project:otari"
    }
  ]
}
```

Return accepted, duplicate, and rejected counts, with bounded per-event validation errors. Re-submitting the same `(source, source_event_id)` must be safe and must not create another row.

The endpoint accepts only metadata and numeric usage fields. It must never accept or persist prompts, completions, tool input, or tool output.

### Persistence and pricing

Extend `usage_logs` with the minimum durable provenance required:

- `source`, non-null, defaulting existing rows to `gateway`
- `source_event_id`, nullable for gateway-originated rows
- optional source attribution suitable for session/project labels
- a uniqueness constraint on `(source, source_event_id)` for imported events

Imported events should use the event timestamp when resolving the effective Otari price. Preserve the existing Anthropic cache accounting semantics:

- normal input is separate from cache-read and cache-write counts
- cache read/write rates apply when configured
- pricing is an API-rate estimate, not an invoice or subscription charge

Expose source and attribution on usage list, summary, CSV, and dashboard activity views. The UI should label external rows and aggregate cost as **estimated API-equivalent cost**, so it cannot be mistaken for spend charged or routed by Otari.

### Budget and mode behavior

Imported usage is observability only:

- do not call reservation, reconciliation, refund, or user-spend mutation paths
- do not affect budget limits or budget usage indicators
- require the master key, like the rest of the usage-management surface
- support standalone mode only
- retain the existing hybrid-mode behavior, where local `/v1/usage` management is unavailable

### Claude Code integration

Document a recommended integration using Claude Code's OpenTelemetry Logs export:

```text
Claude Code OTLP logs
  -> local or hosted OpenTelemetry Collector
  -> normalized external-events batch
  -> Otari
```

The Collector adapter should map Claude Code `api_request` fields directly:

- `request_id` to `source_event_id`
- `event.timestamp` to `timestamp`
- `model`, `input_tokens`, `output_tokens`
- `cache_read_tokens`, `cache_creation_tokens`
- `duration_ms`, `query_source`, and safe session attribution

Do not add a full OTLP receiver to Otari in the first version. OTLP protocol handling, collector durability, batching, and exporter authentication belong in a Collector or sidecar.

For users who cannot enable Claude Code telemetry, document an optional hook-triggered local importer. The hook must only signal an incremental JSONL scanner or sidecar, never synchronously calculate or upload usage in the Claude Code execution path.

## Alternatives considered

- **Route Claude Code through Otari.** This provides normal Otari enforcement and logging, but it uses provider API billing rather than the Claude Code subscription.
- **Parse `ccusage` daily or monthly JSON.** Useful for reporting, but aggregation loses per-event timestamps and durable idempotency.
- **Write directly to Otari's SQLite/PostgreSQL database.** Bypasses authentication, migrations, idempotency, price calculation, and hosted deployments.
- **Add an OTLP receiver directly to Otari.** Larger protocol and operational scope than the first user-facing capability requires.
- **Use a Stop hook as the telemetry source.** Hook payloads do not contain per-request token counts, and the transcript can lag the completed turn.

## Acceptance criteria

- [ ] Standalone Otari accepts bounded batches of normalized external usage events with master-key authentication.
- [ ] Retrying a batch is idempotent by source and source event ID.
- [ ] Imported events are timestamped and priced using Otari's effective configured pricing, including Anthropic cache-read and cache-write pricing.
- [ ] Imported rows never modify `User.spend`, budgets, reservations, or enforcement decisions.
- [ ] Existing gateway rows retain their current behavior and appear as `source = gateway`.
- [ ] Usage list, count, summary, CSV, Activity page, and Usage page expose and clearly label imported usage.
- [ ] Hybrid mode continues not to expose local usage management or ingestion.
- [ ] Integration tests cover auth, validation, idempotency, historical pricing, cache pricing, budget isolation, and migration behavior.
- [ ] Documentation includes a Claude Code OpenTelemetry Collector example and an optional hook-triggered local-importer pattern.
- [ ] OpenAPI and committed dashboard bundle are regenerated.

### PR DESCRIPTION
**Pull request by njbrake**
## Description

Adds an external usage ingestion API so subscription-backed usage (Claude Code first) shows up in Otari at API-equivalent cost, plus a general per-key budget-exemption flag the same usage rides on.

`POST /v1/usage/external-events`: content-free batches (token counts and metadata only; any prompt/completion/tool field is rejected with 422 at both the event and the batch envelope, never stored), idempotent on `(source, source_event_id)`, priced at the effective rate for each event's timestamp (historical pricing, Anthropic cache read/write) or `cost: null` when the model is unpriced. Authenticated with a budget-exempt API key (usage binds to the key's user and stamps its id) or the master key (may name any user). `source` is a generic provenance dimension; `claude_code` is the first source.

`POST /v1/logs` and `POST /v1/traces`: OTLP receivers, so an OpenTelemetry export can point straight at Otari with no collector or sidecar. The logs signal recognizes Claude Code (`api_request`) and Codex (`codex.sse_event` / `codex.api_request`); the traces signal maps any app instrumented with the GenAI semantic conventions (`gen_ai.*`, with `otari.client_name` as provenance). Both protobuf and JSON payloads are accepted, optionally gzipped, with body/event caps and 4xx responses for malformed bodies so exporters never retry-loop. Only content-free usage attributes are read; prompts, tool results, the event's own `cost_usd` (Otari re-prices at its own rate), and user identity are ignored. Usage binds to the import key's user; the master key is refused (it has no user to bind to), and rejected events are reported through the OTLP `partialSuccess` response. Token conventions are normalized per emitter: Claude Code's additive cache counts are booked at the 1-hour write rate, while OpenAI-shaped emitters' cached tokens are de-included from input so cache reads are never double-charged.

General `exclude_from_budget` flag on API keys: imported usage is retrospective and can never be blocked, so an import key must be budget-exempt (a budgeted key is refused with 403). The same flag governs a key's live proxied traffic, which is logged with cost but never reserved, folded into `User.spend`, or gated by `require_pricing`. Implemented by threading a `counts_toward_budget` decision through `ReservationHandle` and gating the spend write in `reconcile_reservation`, so every billable site (chat, messages, responses, passthrough, batches, streaming, vision, attachment top-up) inherits it.

Read surface and dashboard: a `source` filter and a `by_source` breakdown across list, count, summary, and CSV; an `unpriced_requests` count in the summary so a $0 total is not read as free; an API key column on the Activity page, with source and session in the expanded detail; a Tracked cost total (with an unpriced-request count) and an API key filter on the Usage page, keeping Spend for the enforced ledger; an Exempt from budget control (under Advanced) and a Budget-exempt badge on the Keys page. One reversible migration adds the columns and a `(source, source_event_id)` unique constraint; existing rows backfill to `source = gateway`.

Docs: OpenTelemetry is the single documented import path, with dedicated setup guides for Claude Code and Codex (point the telemetry at Otari with an exempt key); `external-usage.md` covers the endpoints, attribute mapping, pricing, and the budget-exempt invariant. Deferred to follow-up issues: per-model and per-request exemption, a subscription-versus-API break-even view, more vendor special-cases, and OTLP over gRPC.

## PR Type

- [x] New Feature
- [ ] Bug Fix
- [ ] Refactor
- [x] Documentation
- [ ] Infrastructure / CI

## Relevant issues

Closes #378

Note: #378 explicitly deferred a full OTLP receiver from the first version. This PR ships one anyway (replacing the issue's collector-plus-importer-script criterion) because it removes the sidecar/collector operational burden and the same mapping generalizes to Codex and any GenAI-instrumented app. Flagging the scope change for the record rather than treating it as incidental.

## Checklist

- [x] I understand the code I am submitting.
- [x] I have added or updated tests that cover my change (`tests/unit`, `tests/integration`).
- [x] I ran the Definition of Done checks locally (`make lint`, `make typecheck`, `make test`).
- [x] Documentation was updated where necessary.
- [x] If the API contract changed, I regenerated the OpenAPI spec (`uv run python scripts/generate_openapi.py`).

## AI Usage

- [ ] No AI was used.
- [ ] AI was used for drafting/refactoring.
- [x] This is fully AI-generated.

**AI Model/Tool used:** Claude Code (Claude Opus 4.8, review hardening by Claude Fable 5)

**Any additional AI details you'd like to share:**

The scope, design decisions, and prioritization are @njbrake's, worked out through back-and-forth; the implementation and prose were written by Claude. The change was driven end to end against a live gateway: real `claude -p` sessions reporting over OpenTelemetry, unpriced turns landing with `cost: null` and priced turns computing full cost including cache, with a non-exempt key correctly refused and `User.spend` unchanged throughout. A six-agent review pass then hardened the OTLP receiver (gzip and size-bound handling, master-key refusal, partialSuccess reporting) and ingestion edge cases, with tests for each fix.

**NOTE:**
When responding to reviewer questions, please respond yourself rather than copy/pasting reviewer comments into an AI and pasting back its answer. We want to discuss with you, not your AI :)

- [x] I am an AI Agent filling out this form (check box if true)

## Summary

- Added an “external usage” import API for subscription-backed usage, with strict batch validation, event deduplication, and attribution to users/sources.
- Added OTLP ingestion endpoints (logs and traces) to import Claude Code, Codex, and GenAI telemetry into the same usage system, supporting JSON/protobuf and gzip, with partial-success reporting for rejected records.
- Introduced a **budget-exempt** option for API keys: imported usage (and exempt-key live gateway traffic) is recorded for analytics but does not reserve budget, affect `User.spend`, enforcement, or pricing gating.
- Expanded usage APIs, CSV exports, and dashboard analytics to filter and break down results by **source** and **API key**, show whether items were priced, and include counts of **unpriced** requests.
- Updated database schema (migration), added integration/unit tests, refreshed documentation and Postman collection, and regenerated dashboard frontend assets.

## Technical notes

- External/OTLP events are content-free and deduplicated by `(source, source_event_id)`.
- Pricing is computed using the configured rate effective at each event’s timestamp (including Anthropic cache read/write pricing); if pricing is missing, cost is left unset and counted as “unpriced.”
- Existing gateway usage is backfilled/kept as `source = gateway`, and imported events cannot claim the `gateway` source.
- Budget-exempt behavior propagates through request settlement and usage logging via a per-record “counts toward budget” flag, ensuring isolation from reservation/spend/enforcement even for live gateway traffic.

### CODE REVIEW

**Comment by njbrake** [comment]
> _Note: this reply was drafted by Claude via back-and-forth with @njbrake. The reasoning and decisions are his; the prose is Claude's._

@coderabbitai thanks for the review. All 10 findings are addressed in cc0f8df3:

- **N+1 (external_usage_service.py):** removed. Users are now validated with a single `IN` query, and pricing is loaded per distinct `(provider, model)` once, then the effective row is resolved in memory per event timestamp. Historical and cache pricing behavior is unchanged (integration tests confirm).
- **Metadata grammar:** `source_event_id`, `provider`, `model`, and `session_label` are constrained to a token-only pattern so prose cannot be smuggled into the content-free fields.
- **Unpriced shown as $0.00:** added `unpriced_requests` to the summary totals; the dashboard Tracked cost tile now shows an "N unpriced" hint so a $0 total is not read as free.
- **Provenance on activity records:** the Source column was intentionally dropped earlier (the API key column carries provenance), so source and session are now shown in the expanded request detail instead.
- **Docs, budget-exempt key:** warns that the flag also exempts live gateway traffic, so it must be treated as a secret and scoped to a dedicated import key/user.
- **Docs, sidecar:** now uses a finite `urlopen` timeout, catches transient errors while preserving unseen events, and chunks batches to the 1000-event cap.
- **Docs, OTel:** specifies `OTEL_LOGS_EXPORTER=otlp` plus endpoint/protocol, since `api_request` rides the logs signal.
- **Tests:** added coverage for the key edit-toggle PATCH (`exclude_from_budget`) and the by_source drill-down navigation.

The one I did not change structurally is the per-group unpriced count in the breakdown; I kept it at the totals level to avoid threading a new field through the top-N fold, which felt heavier than the Minor rating warranted. Happy to reconsider if you disagree.

File: src/gateway/static/dashboard/assets/UsagePage-CR8SlxkC.js [hunk 1]
```diff
@@ -0,0 +1 @@
...y source",rows:(r==null?void 0:r.by_source)??[],totalCost:(s==null?void 0:s.cost)??0,emptyLabel:S?"No usage matches these filters.":"No usage recorded yet.",onDrill:t=>q({source:t}),loading:c.isLoading,columnLabel:"Source",formatKey:t=>t==="gateway"?"Gateway":t==="claude_code"?"Claude Code":t}),e.jsxs("div",{className:"flex flex-col gap-3 rounded-xl border border-[var(--otari-line)] bg-[var(--otari-surface)] p-4",children:[e.jsxs("div",{className:"flex items-center justify-between gap-3",children:[e.jsx("h2",{className:"text-sm font-semibold text-[var(--otari-ink)]",children:"Over time"}),e.jsxs("div",{className:"inline-flex gap-1.5",children:[be.map(t=>e.jsx(y,{size:"sm",variant:v===t.key?"primary":"outline",onPress:()=>d(t.key),children:t.label},t.key)),c.isFetching?e.jsx(V,{size:"sm"}):null]})]}),c.isLoading?e.jsx("div",{className:"flex h-48 items-center justify-center",children:e.jsx(V,{size:"sm"})}):((r==null?void 0:r.series.length)??0)===0?e.jsx("div",{className:"flex h-48 items-center justify-center text-sm text-[var(--otari-muted)]",children:"No data in this range."}):e.jsx(ve,{series:(r==null?void 0:r.series)??[],metric:v,bucket:n.bucket})]})]})]})}export{we as UsagePage};
```
[the remark this answers is not in the document]
Comment (njbrake) [reply src/gateway/static/dashboard/assets/UsagePage-CR8SlxkC.js]: _Note: this comment was drafted by Claude via back-and-forth with @njbrake. The reasoning and decisions are his; the prose is Claude's._

Addressed. Per-row null costs render a dash placeholder rather than $0.00 (Activity rows, request detail, Overview recent activity). The Usage page now labels the total as Tracked cost with an "N unpriced" count beside it, and the summary API exposes unpriced_requests, so a $0 total is never read as free.

**Comment by njbrake** [comment]
As mentioned to Khaled over Slack: merging in order to give time for testing and stabilizition before I head of on vaca.
